### PR TITLE
Stabilize `hint::assert_unchecked`

### PR DESCRIPTION
**Pull request by tgross35**
Make the following API stable, including const:

```rust
// core::hint, std::hint

pub const unsafe fn assert_unchecked(p: bool);
```

This PR also reworks some of the documentation and adds an example.

Tracking issue: https://github.com/rust-lang/rust/issues/119131
FCP: https://github.com/rust-lang/rust/issues/119131#issuecomment-1906394087. The docs update should resolve the remaining concern.

### CODE REVIEW

**Comment by tgross35** [comment]
I don't think there is much purpose in keeping a separate `const_*` feature gate if const stabilization happens at the same time, so merged the gates.

r? libs-api
@rustbot label -T-libs +T-libs-api
cc @scottmcm 
cc @rust-lang/wg-const-eval (and @RalfJung directly since it seems like that ping doesn't always work) for const concerns

**Comment by RalfJung** [comment]
> cc @rust-lang/wg-const-eval (and @RalfJung directly since it seems like that ping doesn't always work) 

It doesn't? I think it always works, except that the person writing the post has to be an org member to ping org teams.

File: library/core/src/hint.rs
Comment (RalfJung) [inline library/core/src/hint.rs:138]: This is potentially misleading. If evaluating the argument has side-effects then it *must* be evaluated in all build modes.

File: library/core/src/hint.rs
Comment (Mark-Simulacrum) [inline library/core/src/hint.rs:158]: This feels like a pretty weird example. Do we not have something better? We even have to defend the example later in the documentation as being not real world.

One example that feels more illustrative is guaranteeing that the returned integer is in-bounds for some slice. For example, [binary search](https://github.com/rust-lang/rust/blob/8f2c255ae03c524323a52d7003ac176beaa3ea74/library/core/src/slice/mod.rs#L2814) tells LLVM that the indexes it returns are in-bounds of the length. Replicating binary search here probably is too much, but replicating a simpler algorithm (e.g., `&[(u32, usize)]` where you search for a particular needle and every usize is guaranteed to be an index in the array -- so before returning it you want to tell LLVM that) feels like a better example.

File: library/core/src/hint.rs
Comment (tgross35) [reply library/core/src/hint.rs]: I updated this to mention side effects

File: library/core/src/hint.rs
Comment (RalfJung) [inline library/core/src/hint.rs:138]:
```suggestion
/// This function produces no code in release mode (but the function argument is
/// still evaluated as usual, just its final value is then ignored).
```

Comment (Mark-Simulacrum) [reply library/core/src/hint.rs:138]: Why is release mode even mentioned here? I feel like that might help clean up dead code, but not otherwise meaningful or guaranteed. For example, I'd expect it to be valid for LLVM to leave everything in if you assert something like is_sorted, just because it doesn't know that function doesn't have side effects (necessarily).

I would just delete this line from the docs.

File: library/core/src/hint.rs
Comment (tgross35) [reply library/core/src/hint.rs:158]: I didn't think there was much purpose in showing a real example of how it may be used in practice since we say elsewhere that you should only use this after benchmarking, and that a smarter optimizer may be able to figure it out at some point. So rather than showing an real usecase that might go away, I figured a dummy example of how it can actually affect codegen may be more useful.

But I am happy to change here as needed, maybe I can rework this example to make its intent more obvious.

**Comment by c410-f3r** [comment]
ping @BurntSushi

Was your concern resolved by the documentation presented in this PR?

File: library/core/src/hint.rs
Comment (BurntSushi) [reply library/core/src/hint.rs:138]: I think I would delete it too.

File: library/core/src/hint.rs
Comment (BurntSushi) [inline library/core/src/hint.rs:140]: Is there a stray backtick here after `If`?

File: library/core/src/hint.rs
Comment (BurntSushi) [reply library/core/src/hint.rs:158]: I like this example. It's not necessarily real world, but I do think it is effective at demonstrating the principle here in a pretty small and easily digestible code snippet.

**Comment by BurntSushi** [comment]
@c410-f3r Yes, I'd be happy to resolve my concern once this is merged.

**Comment by c410-f3r** [comment]
Thanks @BurntSushi

File: library/core/src/hint.rs
Comment (jhpratt) [inline library/core/src/hint.rs:130]:
```suggestion
/// The more complicated the condition, the less likely this is to be useful. For example,
```

For non-native speakers, this is plausibly confusing as written.

**Comment by jhpratt** [comment]
@BurntSushi Given that FCP is blocked on docs and this PR is open to _both_ stabilize and resolve the concern, what would you prefer? Should the PR be split up such that docs can be merged, the concern resolved, and FCP started, or should the concern be lifted on the basis that it _will_ be resolved at the same time as stabilization? I'd prefer the latter for simplicity.

**Comment by tgross35** [comment]
Ah this slipped my mind but is waiting on me, I’ll incorporate the feedback tomorrow. I think burntsushi said somehow that this PR is suitable to resolve the docs concern.

@rustbot author

**Comment by BurntSushi** [comment]
Yeah I see the conflict here. I think when I was reviewing this, I thought this was just a doc update. Yes, in the future, I think that should be de-coupled in the FCP, otherwise we have a chicken & egg problem. Generally speaking, I'd like to see doc updates merged before resolving my concern. But I've resolved the concern in #119131.

File: library/core/src/hint.rs
Comment (tgross35) [reply library/core/src/hint.rs]: Thanks, applied

File: library/core/src/hint.rs
Comment (tgross35) [reply library/core/src/hint.rs]: Thanks, fixed it

File: library/core/src/hint.rs
Comment (tgross35) [reply library/core/src/hint.rs]: Removed this line

**Comment by tgross35** [comment]
I updated the docs based on the feedback. Since you have been pretty involved here,
r? @BurntSushi

**Comment by tgross35** [comment]
Oh whoops, forgot the important labels

@rustbot ready

**Review by dtolnay — APPROVED** [review]
Thanks!

**Pull request merged**
